src/sql/src/query_model/mod.rs

```diff
@@ -0,0 +1,417 @@
+                }
+            }
+            BoxType::Except | BoxType::Union | BoxType::Intersect | BoxType::BaseTable(_) => {}
+        }
+        Ok(())
+    }
+
+    fn is_select(&self) -> bool {
+        matches!(self.box_type, BoxType::Select(_))
+    }
+}
+
+#[derive(Debug)]
+pub enum BoxType {
+    BaseTable(BaseTable),
+    Except,
+    Grouping(Grouping),
+    Intersect,
+    OuterJoin(OuterJoin),
+    Select(Select),
+    TableFunction(TableFunction),
+    Union,
+    Values(Values),
+}
+
+impl BoxType {
+    fn get_box_type_str(&self) -> &'static str {
+        match self {
+            BoxType::BaseTable(..) => "BaseTable",
+            BoxType::Except => "Except",
+            BoxType::Grouping(..) => "Grouping",
+            BoxType::Intersect => "Intersect",
+            BoxType::OuterJoin(..) => "OuterJoin",
+            BoxType::Select(..) => "Select",
+            BoxType::TableFunction(..) => "TableFunction",
+            BoxType::Union => "Union",
+            BoxType::Values(..) => "Values",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum DistinctOperation {
+    Enforce,
+    Guaranteed,
+    Preserve,
+}
+
+pub use sql_parser::ast::Ident;
+
+#[derive(Debug)]
+pub struct Quantifier {
+    /// uniquely identifiers the quantifier within the model
+    pub id: QuantifierId,
+    /// the type of the quantifier
+    pub quantifier_type: QuantifierType,
+    /// the input box of this quantifier
+    pub input_box: BoxId,
+    /// the box that owns this quantifier
+    pub parent_box: BoxId,
+    /// alias for name resolution purposes
+    pub alias: Option<Ident>,
+}
+
+#[derive(Debug)]
+pub enum QuantifierType {
+    All,
+    Existential,
+    Foreach,
+    PreservedForeach,
+    Scalar,
+}
+
+impl QuantifierType {
+    fn is_subquery(&self) -> bool {
+        match self {
+            QuantifierType::All | QuantifierType::Existential | QuantifierType::Scalar => true,
+            _ => false,
+        }
+    }
+}
+
+impl fmt::Display for QuantifierType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self {
+            QuantifierType::Foreach => write!(f, "F"),
+            QuantifierType::PreservedForeach => write!(f, "P"),
+            QuantifierType::Existential => write!(f, "E"),
+            QuantifierType::All => write!(f, "A"),
+            QuantifierType::Scalar => write!(f, "S"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct BaseTable {/* @todo table metadata from the catalog */}
+
+#[derive(Debug)]
+pub struct Grouping {
+    pub key: Vec<Box<Expr>>,
+}
+
+#[derive(Debug)]
+pub struct OuterJoin {
+    pub predicates: Vec<Box<Expr>>,
+}
+
+impl OuterJoin {
+    #[allow(dead_code)]
+    fn new() -> Self {
+        Self {
+            predicates: Vec::new(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Select {
+    pub predicates: Vec<Box<Expr>>,
+    pub order_key: Option<Vec<Box<Expr>>>,
+    pub limit: Option<Expr>,
+    pub offset: Option<Expr>,
```

Comment:
Can `limit` and `offset` ever be anything beside literal uints?

Reply:
Not sure about the standard, but at least Postgres supports anything that is a constant within the context of the query block, for example, lateral column references.

```
asenac=> select * from t1, lateral(select * from t2 limit t1.f1 offset t1.f2) foo;
 f1 | f2 | f1 | f2 
----+----+----+----
(0 rows)

asenac=>
```

Reply:
Why is translating HirRelationExpr::TopK to QGM currently an open question according to #8073?

Hmm...neither `HirRelationExpr` nor `MirRelationExpr` supports `limit` or `offset` being anything besides literal uints. Thus, I lean towards having `limit` and `offset` be uints until 1) we decide to work on adding support for generalized `limit`/`offset` to the public interface or 2) we have a DSL for creating a QGM directly so we can test the generalized `limit`/`offset`.

Reply:
Because `TopK` is the combination QGM's Grouping and Select operator in a single operator, so I guess the only option we have is to add a TopK box in QGM.

Reply:
BTW, there is an issue tracking adding support for non-literal limit/offset https://github.com/MaterializeInc/materialize/issues/7071